nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -146,24 +147,44 @@ impl DataStore {
             })
     }
 
-    pub async fn volume_hard_delete(&self, volume_id: Uuid) -> DeleteResult {
+    /// Return a `Option<Volume>` based on id, even if it's soft deleted.
+    pub async fn volume_get(
+        &self,
+        volume_id: Uuid,
+    ) -> LookupResult<Option<Volume>> {
         use db::schema::volume::dsl;
-
-        diesel::delete(dsl::volume)
+        dsl::volume
             .filter(dsl::id.eq(volume_id))
-            .execute_async(self.pool())
+            .select(Volume::as_select())
+            .first_async::<Volume>(self.pool())
+            .await
+            .optional()
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
+    }
+
+    /// Delete the volume if it exists. If it was already deleted, this is a
+    /// no-op.
+    pub async fn volume_hard_delete(&self, volume_id: Uuid) -> DeleteResult {
+        self.pool()
+            .transaction(move |conn| {
+                use db::schema::volume::dsl;
+
+                let volume = dsl::volume
+                    .filter(dsl::id.eq(volume_id))
+                    .select(Volume::as_select())
+                    .first::<Volume>(conn)
+                    .optional()?;
+
+                if volume.is_some() {
+                    diesel::delete(dsl::volume)
+                        .filter(dsl::id.eq(volume_id))
+                        .execute(conn)?;
+                }
```

Comment:
Do we need to do the lookup first?

If we just unconditionally performed:

```rust
// We don't care if we deleted this volume, or if it has already been deleted.
let _row_count = diesel::delete(dsl::volume).filter(dsl::id.eq(volume_id)).execute(conn)?;
Ok(())
```

This seems like it would use one less query, and wouldn't require a transaction.

Reply:
I did this for two reasons:

1) to separate out the errors: in the case where we have an explicit check then delete, any error we receive from the delete is unrelated to the record missing. Although, now that I type this out: what other error could we receive from a delete call?

2) it's more clearly idempotent to my eyes, although I'll admit that's a weaker justification.

Reply:
Do we actually do (1) though? If no such volume exists, we don't throw an error, do we? Don't we just set `let volume = None`, and then skip the delete?

The cases of "the volume exists / doesn't exist" seem the same to me, but doing the deletion with these two steps just requires an extra query

Reply:
You're right, I've changed to just the delete in 9efe842a4. `test_volume_hard_delete_idempotent` would have caught this if it wasn't true!